runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1174,23 +1029,400 @@ static iree_status_t iree_hal_hip_device_queue_write(
     iree_hal_buffer_t* source_buffer, iree_device_size_t source_offset,
     iree_hal_file_t* target_file, uint64_t target_offset,
     iree_device_size_t length, iree_hal_write_flags_t flags) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
   // TODO: expose streaming chunk count/size options.
   iree_status_t loop_status = iree_ok_status();
   iree_hal_file_transfer_options_t options = {
       .loop = iree_loop_inline(&loop_status),
       .chunk_count = IREE_HAL_FILE_TRANSFER_CHUNK_COUNT_DEFAULT,
       .chunk_size = IREE_HAL_FILE_TRANSFER_CHUNK_SIZE_DEFAULT,
   };
-  IREE_RETURN_IF_ERROR(iree_hal_device_queue_write_streaming(
-      base_device, queue_affinity, wait_semaphore_list, signal_semaphore_list,
-      source_buffer, source_offset, target_file, target_offset, length, flags,
-      options));
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_device_queue_write_streaming(
+              base_device, queue_affinity, wait_semaphore_list,
+              signal_semaphore_list, source_buffer, source_offset, target_file,
+              target_offset, length, flags, options));
+  IREE_TRACE_ZONE_END(z0);
   return loop_status;
 }
 
-static void iree_hal_hip_device_collect_tracing_context(void* user_data) {
-  iree_hal_stream_tracing_context_collect(
-      (iree_hal_stream_tracing_context_t*)user_data);
+typedef struct iree_hal_hip_device_semaphore_submit_callback_data_t {
+  iree_atomic_ref_count_t wait_semaphore_count;
+  iree_hal_hip_device_t* device;
+  iree_hal_queue_affinity_t queue_affinity;
+  iree_hal_command_buffer_t* command_buffer;
+  iree_hal_buffer_binding_table_t binding_table;
+  iree_hal_semaphore_list_t wait_semaphore_list;
+  iree_hal_semaphore_list_t signal_semaphore_list;
+  iree_hal_resource_set_t* resource_set;
+  iree_slim_mutex_t status_mutex;
+  iree_status_t status;
+} iree_hal_hip_device_semaphore_submit_callback_data_t;
+
+static iree_status_t iree_hal_hip_device_complete_submission(
+    void* user_data, iree_hal_hip_event_t* event, iree_status_t status) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_hal_hip_device_semaphore_submit_callback_data_t* data =
+      (iree_hal_hip_device_semaphore_submit_callback_data_t*)user_data;
+  iree_hal_hip_device_t* device = data->device;
+
+  // Get the device_context from the queue_affinity.
+  uint32_t idx = iree_math_count_trailing_zeros_u64(data->queue_affinity);
+
+  // 1) Read any tracing events that were submitted.
```

Comment:
the comments are useful but numbers like this bitrot fast as code is refactored/inserted/removed - don't use them inline and if you think it'd help clarify the algorithm to readers then include an implementation comment that walks through the algorithm. if there's things that need to happen in a specific order that's not obvious (because it happens on the other side of a function call or something) then you can call that out instead of making the user scan around for a numbered comment.

```suggestion
  // Read any tracing events that were submitted.
  
```